### PR TITLE
Move to starknet.py 0.16.1

### LINKED ISSUE #2043: Bump starknet-py from 0.15.2 to 0.16.1
https://github.com/software-mansion/protostar/pull/2038

### PR DESCRIPTION
**Pull request by Radinyn**
Moved to starknet.py 0.16.1, includes many bundling changes along a reworked data transformer. One of the safety features (checking if a user passes a dict and not a felt pointer to a function) was lost.

### CODE REVIEW

**Comment by Radinyn** [comment]
Removing the DataTransformer entirely might not be the best solution, I'll consider recreating it with starknet.py components.

File: poetry.lock [hunk 1]
```diff
@@ -1,100 +1,10 @@
-# This file is automatically @generated by Poetry 1.5.1 and should not be changed by hand.
-
 [[package]]
 name = "aiohttp"
 version = "3.8.4"
 description = "Async http client/server framework (asyncio)"
+category = "main"
```
Comment (abulenok) [inline poetry.lock:5]: Are changes to the `.lock` file intended?

File: poetry.lock [hunk 2]
```diff
@@ -1,100 +1,10 @@
...
 optional = false
 python-versions = ">=3.6"
-files = [
-    {file = "aiohttp-3.8.4-cp310-cp310-macosx_10_9_universal2.whl", hash = "sha256:5ce45967538fb747370308d3145aa68a074bdecb4f3a300869590f725ced69c1"},
-    {file = "aiohttp-3.8.4-cp310-cp310-macosx_10_9_x86_64.whl", hash = "sha256:b744c33b6f14ca26b7544e8d8aadff6b765a80ad6164fb1a430bbadd593dfb1a"},
-    {file = "aiohttp-3.8.4-cp310-cp310-macosx_11_0_arm64.whl", hash = "sha256:1a45865451439eb320784918617ba54b7a377e3501fb70402ab84d38c2cd891b"},
-    {file = "aiohttp-3.8.4-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl", hash = "sha256:a86d42d7cba1cec432d47ab13b6637bee393a10f664c425ea7b305d1301ca1a3"},
-    {file = "aiohttp-3.8.4-cp310-cp310-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl", hash = "sha256:ee3c36df21b5714d49fc4580247947aa64bcbe2939d1b77b4c8dcb8f6c9faecc"},
-    {file = "aiohttp-3.8.4-cp310-cp310-manylinux_2_17_s390x.manylinux2014_s390x.whl", hash = "sha256:176a64b24c0935869d5bbc4c96e82f89f643bcdf08ec947701b9dbb3c956b7dd"},
-    {file = "aiohttp-3.8.4-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:c844fd628851c0bc309f3c801b3a3d58ce430b2ce5b359cd918a5a76d0b20cb5"},
```
Comment (abulenok) [inline poetry.lock:16]: Are these changes intended?

Comment (abulenok) [reply poetry.lock:16]: It seems that you are using older version of poetry

https://github.com/orgs/python-poetry/discussions/7445#discussioncomment-4832320
https://github.com/orgs/python-poetry/discussions/6763

File: protostar/git/git_helpers.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def run_git(*args: Union[str, Path], cwd: Path):
 
 def has_user_git_credentials():
     try:
-        subprocess.run(["git", "config", "user.name"], check=True, **SHARED_KWARGS)
+        subprocess.run(["git", "config", "user.name"], check=True, **SHARED_KWARGS)  # type: ignore
```
Comment (abulenok) [inline protostar/git/git_helpers.py:53]: why do we need this?

File: tests/_conftest/compiled_account/compiler.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
+"""
+DEPRECATED, will be removed soon
```
Comment (abulenok) [inline tests/_conftest/compiled_account/compiler.py:2]: Do we have to remove this? Is there an issue for it?

Comment (Radinyn) [reply tests/_conftest/compiled_account/compiler.py:2]: Whole Cairo0 is deprecated, so it just won't be used soon.

File: protostar/git/git_helpers.py (hunk 3 above)
Comment (Radinyn) [reply protostar/git/git_helpers.py:53]: pyright suddenly exploded with 170 lines of type errors here

Comment (Radinyn) [reply protostar/git/git_helpers.py:53]: [same problem](https://app.circleci.com/pipelines/github/software-mansion/protostar/777/workflows/b07068d9-dcb8-4a7f-b9b9-c705e70c74b1/jobs/5215)

**Review by Arcticae — APPROVED** [review]
